### PR TITLE
Add support for hot module reload.

### LINKED ISSUE #14: Hot module reload
I've setup hot module reload on a project and it's working nicely, except when adding/removing/editing svgs that go through this loader.

Maybe that's something that we could improve? See style-loader for an example: https://github.com/webpack-contrib/style-loader/blob/master/index.js#L21

### PR DESCRIPTION
**Pull request by satazor**
This commit adds hot module reload support:
- When [hash] is used, the previous hashed URLs will be correctly replaced with the new ones
- When [hash] is not used, a cache-bust will be added to the URL to force browsers to re-download the sprite; this only happens on the first render so that it plays well with server-side rendering, otherwise many view libraries will complain about mismatches during rehydration (such as React).

Moreover, a few more changes were done:
- Removed `cacheable(false)` because it's no longer necessary and actually was hurting performance a lot
- Fix dots and other chars not being correctly escaped in regular expressions by using lodash
- Improve performance by not doing unnecessary work, such as replacements, if the sprite hasn't changed
- Removed unused `resourcePath` from SvgIcon
- Added support for HRM with react-hot-loader in the example app, with and without [hash]
- Changed external-svg-sprite-loader imports in the example app to be relative so that we can test changes in the library without having to re-install the dependencies

Closes #14.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -27,27 +28,30 @@ const DEFAULT_QUERY_VALUES = {
...
+    const { addDependency, resourcePath, _compiler } = this;
 
     // Get callback because the SVG is going to be optimized and that is an async operation
     const callback = this.async();
 
     // Parse the loader query and apply the default values in case no values are provided
     const query = Object.assign({}, DEFAULT_QUERY_VALUES, loaderUtils.getOptions(this));
 
+    // Check if HMR is enabled
+    const isHMREnabled = !!_compiler.options.plugins.find((plugin) => plugin.constructor.name === 'HotModuleReplacementPlugin');
```
Comment (satazor) [inline index.js]: Let me know if this verification is ok or if you want this under an option, e.g.: `loader: 'external-svg-sprite-loader?hot'`.

**Comment by satazor** [comment]
Please note that the CSS is not hot module reloading because https://github.com/webpack-contrib/mini-css-extract-plugin still doesn't support it. There's an open MR though, I will test it in a few minutes: https://github.com/webpack-contrib/mini-css-extract-plugin/pull/100.

File: index.js (hunk 1 above)
Comment (satazor) [reply index.js]: I've actually replaced it with the `hot` option. The reason is that accessing `_compiler` on `this` is deprecated.

File: index.js [hunk 2]
```diff
@@ -64,20 +62,48 @@ function loader(content) {
...
-                        symbol: publicPath + '${icon.getUrlToSymbol()}',
-                        view: publicPath + '${icon.getUrlToView()}',
-                        viewBox: '${icon.getDocument().getViewBox()}',
-                        title: '${icon.getDocument().getTitle()}',
-                        toString: function () {
-                            return JSON.stringify(this.view);
-                        }
-                    };`
-                );
+                const code = query.hot ?
```
Comment (satazor) [inline index.js]: I could have placed everything into the same string, but every icon would have all this new code in production, which is unnecessary.

**Comment by satazor** [comment]
There's one additional performance improvement we can make. At the moment, the sprite is being generated and emitted everytime a recompilation occurs, even if the sprite hasn't changed. This usually doesn't take much time but it increases linearly with the amount of icons in the application.

Want me to create a MR that attempts to improve on that?

File: index.js [hunk 3]
```diff
@@ -64,20 +62,48 @@ function loader(content) {
...
+                    dedent(`
+                        var publicPath = ${query.publicPath ? `'${query.publicPath}'` : '__webpack_public_path__'};
+                        var isDocumentReady = typeof document !== 'undefined' ? document.readyState === 'complete' : false;
+                        var maybeAddCacheBustingToUrl = isDocumentReady && module.hot ?
```
Comment (PedroMiguelSS) [inline index.js]: I have a question @satazor. Can you explain what is the difference between `query.hot` and `module.hot`?

Comment (satazor) [reply index.js]: module.hot is a way to detect hot module replacement at runtime.

query.hot is an option to enable hot module replacement via options at setup time. The benefit is that all this new code to support hot replacement will not be added in production as well as users are able to opt-out/in, even if HMR is enabled

Comment (PedroMiguelSS) [reply index.js]: Ok, that's what I was supposing. Thanks!

**Comment by satazor** [comment]
Mentioning @bensampaio just to get some feedback on this :D

File: index.js (hunk 2 above)
Comment (bensampaio) [inline index.js]: There is a lot of repetition here. This really increases the maintainability costs of this plugin since the output is repeated twice. But before I can suggest a solution I need you to answer the other comments.

File: index.js [hunk 4]
```diff
@@ -64,20 +62,48 @@ function loader(content) {
...
+                            function (url) {
+                                var parts = url.split('#');
+
+                                return parts[0] + '?' + parts[1] + '#' + parts[1];
```
Comment (bensampaio) [inline index.js]: Why is this necessary? What's happening here? Is this a common practice when using HMR?

File: index.js [hunk 5]
```diff
@@ -64,20 +62,48 @@ function loader(content) {
...
-                        title: '${icon.getDocument().getTitle()}',
-                        toString: function () {
-                            return JSON.stringify(this.view);
-                        }
-                    };`
-                );
+                const code = query.hot ?
+                    dedent(`
+                        var publicPath = ${query.publicPath ? `'${query.publicPath}'` : '__webpack_public_path__'};
+                        var isDocumentReady = typeof document !== 'undefined' ? document.readyState === 'complete' : false;
```
Comment (bensampaio) [inline index.js]: Why do you need this check?

File: index.js (hunk 4 above)
Comment (satazor) [reply index.js]: It's not a common practice, usually loaders use `module.hot.accept` to provide HMR. Let me try to explain why we need this:

In the default configuration, which is suitable for development, `options.name` is `img/sprite.svg`. The side effect of this is that the sprite URL will be the same. This causes the following issues:

- Let's imagine a react component that uses the `symbol` to generate a `<svg>` tag. Because the URL hasn't changed, react will not do any DOM operation. Because of this, the updated sprite will not be fetched.
- A CSS file has the same issue as the browser recomputes the styles and sees no difference.
- Even if the two prior points were not a problem, the browser might not re-download the sprite if the server didn't set the right headers for development

Because of this, I've come up with the idea of using cache-busting to get around the issue as anyone will see the sprite URL as a new URL.

File: index.js (hunk 2 above)
Comment (satazor) [reply index.js]: I agree. The repetition was made on purpose so that production code doesn't include logic only necessary to development.

File: index.js (hunk 5 above)
Comment (bensampaio) [reply index.js]: Is cache busting really necessary when you use a hash in the sprite name? I rather make that mandatory then use cache busting.

File: index.js (hunk 4 above)
Comment (bensampaio) [reply index.js]: Is cache busting really necessary when you use a hash in the sprite name? I rather make that mandatory then use cache busting.

File: package.json [hunk 6]
```diff
@@ -13,6 +13,7 @@
   ],
   "dependencies": {
     "cheerio": "^0.20.0",
+    "dedent": "^0.7.0",
```
Comment (bensampaio) [inline package.json]: Why add this as a dependency? When you uglify the code it will be all in one line anyway. The spacing shouldn't matter.

File: examples/react/webpack.config.js [hunk 7]
```diff
@@ -29,9 +29,10 @@ module.exports = {
                 test: /\.css$/,
             },
             {
-                loader: 'external-svg-sprite-loader',
+                loader: require.resolve('../..'),
                 options: {
-                    name: 'img/sprite.[hash].svg',
+                    name: 'img/sprite.svg',
```
Comment (bensampaio) [inline examples/react/webpack.config.js]: I want these example to work with the `[hash]` so that shouldn't be removed.

**Comment by bensampaio** [comment]
@satazor thanks for the detailed answers. I left you some more questions. I will try to focus on this for the coming weeks since I would like to release v4 soon and I would like to include these changes on it.

File: index.js (hunk 4 above)
Comment (satazor) [reply index.js]: Using a hash in the sprite name in development is a bad practice. The webpack dev server uses in-memory filesystem to improve performance. Using a hash will fill the in-memory file-system.

File: index.js (hunk 5 above)
Comment (satazor) [reply index.js]: Answered above.

File: package.json (hunk 6 above)
Comment (satazor) [reply package.json]: Will remove.

File: examples/react/webpack.config.js (hunk 7 above)
Comment (bensampaio) [reply examples/react/webpack.config.js]: Ok, I didn't know using a hash in dev is considered a bad practice. In that case, I think we can base it on the environment? Like `process.env.NODE_ENV === 'production'? 'img/sprite.[hash].svg' : 'img/sprite.svg'`. What do you think?

Comment (satazor) [reply examples/react/webpack.config.js]: That would work but it's kind of strange to have `NODE_ENV=production npm start` where `npm start` runs `webpack-dev-server`, which is.. well.. a development server.

`webpack-dev-server` has a `--hot` flag to enable HRM. What if we detect the presence of that flag instead? (by using https://github.com/sindresorhus/has-flag)

File: index.js [hunk 8]
```diff
@@ -64,10 +61,35 @@ function loader(content) {
...
+                    var maybeAddCacheBustingToUrl = isDocumentReady && module.hot ?
+                        function (url) {
+                            var parts = url.split('#');
+
+                            return parts[0] + '?' + parts[1] + '#' + parts[1];
+                        } :
+                        function (url) { return url; };
+
+                    module.exports = {
+                        symbol: publicPath + maybeAddCacheBustingToUrl('${icon.getUrlToSymbol()}'),
```
Comment (bensampaio) [inline index.js]: Would it be possible to pass the value of `query.hot` into `getUrlToSymbol ` and `getUrlToView`, so that if true they both add the cache busting param? In this way would be possible to reduce the repetition of the output block. Something like:

```
`var publicPath = ${query.publicPath ? `'${query.publicPath}'` : '__webpack_public_path__'};
module.exports = {
    symbol: publicPath + '${icon.getUrlToSymbol(query.hot)}',
    view: publicPath + '${icon.getUrlToView(query.hot)}',
    viewBox: '${icon.getDocument().getViewBox()}',
    title: '${icon.getDocument().getTitle()}',
    toString: function () {
        return JSON.stringify(this.view);
    }
};`
```

```
getUrlToSymbol(hot) {
    const { sprite, symbolName } = this;

    const query = hot ? '?' + symbolName : '';
    const hash = '#' + symbolName;

    return sprite.resourcePath + query + hash;
}

getUrlToView(hot) {
    const { sprite, viewName } = this;

    const query = hot ? '?' + viewName : '';
    const hash = '#' + viewName;

    return sprite.resourcePath + query + hash;
}
```

This would be the ideal solution imo. But the "document is ready" part is really necessary then I suggest the following:

```
`var publicPath = ${query.publicPath ? `'${query.publicPath}'` : '__webpack_public_path__'};
var symbolUrl = '${icon.getUrlToSymbol()}';
var viewUrl = '${icon.getUrlToView()}';
if (${query.hot})
	var isDocumentReady = typeof document !== 'undefined' ? document.readyState === 'complete' : false;
	function maybeAddCacheBustingToUrl(url) {
	    var parts = url.split('#');

	    if (isDocumentReady && module.hot) {
	    	return parts[0] + '?' + parts[1] + '#' + parts[1];
	    }

	    return url;
	}

	symbolUrl = maybeAddCacheBustingToUrl(symbolUrl);
	viewUrl = maybeAddCacheBustingToUrl(viewUrl);
}
module.exports = {
    symbol: publicPath + symbolUrl,
    view: publicPath + viewUrl,
    viewBox: '${icon.getDocument().getViewBox()}',
    title: '${icon.getDocument().getTitle()}',
    toString: function () {
        return JSON.stringify(this.view);
    }
};`
```

In this last option, webpack would remove the entire if block if `query.hot` evaluates to false. in this way, the `module.exports` section would not be touched (which is what I want to accomplish) and the extra code would still not be present when HMR is not on.

What do you think?

Comment (satazor) [reply index.js]: I think that the last option should work and it's necessary to have the document.ready thing. Could you test it out? I'm currently very busy due to a tight deadline on a project and would love some help :D

File: examples/react/webpack.config.js (hunk 7 above)
Comment (bensampaio) [reply examples/react/webpack.config.js]: I prefer not to add an extra dependency for this. I think we can instead change the npm scripts to:

```
"start:dev": "webpack-dev-server --content-base public/ --port 3000",
"build:prd": "NODE_ENV=production webpack -p --progress"
```

Then we can easily start the server for development or build the assets for production and also check if the prd output is the expected one. For this to work, when `NODE_ENV === production` the `mode` in `webpack.config.js` needs to be set to `production`.

Comment (satazor) [reply examples/react/webpack.config.js]: Alright. Just for reference, this dependency would just be on the example and not on the actual source code.

Comment (satazor) [reply examples/react/webpack.config.js]: Just updated the above comment.

File: index.js (hunk 8 above)
Comment (bensampaio) [reply index.js]: I would like to test it but I am also busy with a deadline and I am not working on any project with HMR atm. But I assume you are, or is that not the case? I could test it with the example projects but I am not sure if that would be enough. If you don't find the time in the coming weeks it's ok.

Comment (satazor) [reply index.js]: Yes I'm working on a project with HRM but restarting the server whenever the icons are updated or added isn't really an issue. To be honest, the biggest benefit of this MR was to enable caching which improved the HMR times a LOT.

I will test it out when I got sometime. Feel free to keep pinging me if I don't respond.
@andregoncalvesdev would you have the time to test @bensampaio suggestion?

File: examples/react/webpack.config.js (hunk 7 above)
Comment (bensampaio) [reply examples/react/webpack.config.js]: Yes, let's do that. I know the dependency would only be in the examples but I intend to minimize the amount of dependencies as much as possible because in the end I still have to maintain them.

File: index.js (hunk 8 above)
Comment (andregoncalvesdev) [reply index.js]: @satazor i´ll try to get time to test it at the eow.

Comment (usebaz) [reply index.js]: hi, any updates?
@andregoncalvesdev

Comment (satazor) [reply index.js]: @usebaz @andregoncalvesdev has been unable to tackle this due to lack of time. Will try to do it tonight or in the next few days as I will be on holidays and will have more time.

Comment (bensampaio) [reply index.js]: @satazor any progress?

File: index.js [hunk 9]
```diff
@@ -64,13 +61,33 @@ function loader(content) {
...
+            // It's important that the cache busting is not included initially so that it plays well with server-side rendering,
+            // otherwise many view libraries will complain about mismatches during rehydration (such as React)
             setImmediate(() => {
                 callback(
                     null,
                     `var publicPath = ${query.publicPath ? `'${query.publicPath}'` : '__webpack_public_path__'};
+                    var symbolUrl = '${icon.getUrlToSymbol()}';
+                    var viewUrl = '${icon.getUrlToView()}';
+
+                    if (${!!query.hot}) {
```
Comment (satazor) [inline index.js]: When doing a production build, it fails on the uglify: 

```
    ERROR in js/about.214688d4aa3a710.js from UglifyJs
    Unexpected token: punc
```

will need to investigate.

**Comment by satazor** [comment]
@bensampaio @usebaz I've updated the PR, it contains substancial changes from the last one. See my updated commit message or the message here: https://github.com/karify/external-svg-sprite-loader/pull/47#issue-184782440

**Comment by satazor** [comment]
Please review the changes carefully and give me feedback!

File: lib/SvgSprite.js [hunk 10]
```diff
@@ -86,6 +81,11 @@ class SvgSprite {
         deltaY = 0,
         iconHeight = DEFAULT_ICON_HEIGHT,
     } = {}) {
+        if (!this.dirty) {
+            this.updated = false;
```
Comment (bensampaio) [inline lib/SvgSprite.js]: You should define the default value for this property in the constructor. Adding properties to classes that were not defined in the constructor might cause a performance penalty.

File: lib/SvgSprite.js [hunk 11]
```diff
@@ -154,14 +154,16 @@ class SvgSprite {
         );
 
         // Generate interpolated name
-        const resourcePath = loaderUtils.interpolateName({}, this.originalPath, { content });
+        const resourcePath = loaderUtils.interpolateName({}, this.originalResourcePath, { content });
 
         // Assign resource path and content to public values
+        this.previousResourcePath = this.resourcePath;
```
Comment (bensampaio) [inline lib/SvgSprite.js]: You should define the default value for this property in the constructor. Adding properties to classes that were not defined in the constructor might cause a performance penalty.

File: lib/SvgStorePlugin.js [hunk 12]
```diff
@@ -150,25 +152,39 @@ class SvgStorePlugin {
      * @param {SvgSprite} sprite - the sprite for the module.
      */
     replaceSpritePathInModuleSource(module, sprite) {
-        const { originalPathRegExp, resourcePath } = sprite;
+        const { originalResourcePath, previousResourcePath, resourcePath } = sprite;
+
+        // Always replace the `originalResourcePath`
+        // Moreover, replace the `previousResourcePath` with the new one if they are different
+        const originalResourcePathRegExp = new RegExp(escapeRegExp(originalResourcePath), 'gm');
```
Comment (bensampaio) [inline lib/SvgStorePlugin.js]: There is too much happening in this line. Can you move `escapeRegExp(originalResourcePath)` to a separate const?

File: lib/SvgStorePlugin.js [hunk 13]
```diff
@@ -150,25 +152,39 @@ class SvgStorePlugin {
...
+        const previousResourcePathRegExp = resourcePath !== previousResourcePath ? new RegExp(escapeRegExp(previousResourcePath), 'gm') : null;
```
Comment (bensampaio) [inline lib/SvgStorePlugin.js]: Can you also move `escapeRegExp(previousResourcePath)` to a separate const?

File: lib/SvgStorePlugin.js [hunk 14]
```diff
@@ -150,25 +152,39 @@ class SvgStorePlugin {
...
+        const replacePaths = (input) => {
```
Comment (bensampaio) [inline lib/SvgStorePlugin.js]: `replaceSpritePathInModuleSource` is called multiple times so creating a function inside of this method isn't really a good idea for performance. I suggest renaming this method to `replaceSpritePathsForModuleType` and then create a new method `replaceSpritePathsInModuleSource` like:

```
replaceSpritePathsInModuleSource(source, resourcePath, originalResourcePathRegExp, previousResourcePathRegExp) {
  source = source.replace(originalResourcePathRegExp, resourcePath);

  if (previousResourcePathRegExp) {
    source = source.replace(previousResourcePathRegExp, resourcePath);
  }

  return source;
}
```

Comment (satazor) [reply lib/SvgStorePlugin.js]: Nice catch. Even if I do that, the regular expressions will still be created several times for each chunk and module. What if I move the function to the SvgSprite instead and only create the regular expressions whenever the sprite was updated?

File: lib/SvgSprite.js (hunk 10 above)
Comment (satazor) [reply lib/SvgSprite.js]: Done.

File: lib/SvgSprite.js (hunk 11 above)
Comment (satazor) [reply lib/SvgSprite.js]: Done.

File: lib/SvgStorePlugin.js (hunk 14 above)
Comment (bensampaio) [reply lib/SvgStorePlugin.js]: Yeah, I think that would work. Good idea.

Comment (satazor) [reply lib/SvgStorePlugin.js]: Done.

File: lib/SvgStorePlugin.js (hunk 13 above)
Comment (satazor) [reply lib/SvgStorePlugin.js]: Can you check the changes and let me know if you still think we should put it in another line?

File: lib/SvgStorePlugin.js (hunk 12 above)
Comment (satazor) [reply lib/SvgStorePlugin.js]: Can you check the changes and let me know if you still think we should put it in another line?

File: lib/SvgSprite.js [hunk 15]
```diff
@@ -154,14 +160,37 @@ class SvgSprite {
...
+        this.previousResourcePathRegExp = this.resourcePathRegExp;
```
Comment (bensampaio) [inline lib/SvgSprite.js]: This should also be set in the constructor

Comment (bensampaio) [reply lib/SvgSprite.js]: @satazor if you change this I would say this is ready for merge :)

Comment (satazor) [reply lib/SvgSprite.js]: Done!

Could you please merge after lunch? I would like to some tests after PR review changes on https://github.com/moxystudio/react-with-moxy and post some screenshots about the performance differences as well!

**Comment by satazor** [comment]
Everything is working as expected in my tests. Here's the performance results in our [react boilerplate](https://github.com/moxystudio/react-with-moxy), using [speed-measure-webpack-plugin](https://github.com/stephencookdev/speed-measure-webpack-plugin):

**Initial compilation**:

screen shot 2018-08-24 at 15 11 47

Outcome: similar results, as expected

**Editing one SVG**:

screen shot 2018-08-24 at 15 12 10

Outcome: from 84 svgs processed to just 1; this was due to the removal of `cacheable(false)`

**Doing a change in a react component**:

screen shot 2018-08-24 at 15 20 58

Outcome: now, the the loader and the plugin aren't doing anything.. previously they were doing a lot of unnecessary work

@bensampaio Hope you like it, feel free to merge!

File: lib/SvgSprite.js (hunk 15 above)
Comment (bensampaio) [reply lib/SvgSprite.js]: Sorry I couldn't do this earlier. Had a pretty busy day.

**Comment by bensampaio** [comment]
@satazor awesome work. Hopefully next week I will have time to polish some stuff up and release a new alpha version. And then a few weeks later I will release v4.

**Comment by satazor** [comment]
An alpha release next week would be awesome, looking forward to it!

**Comment by bensampaio** [comment]
Done, 4.0.0-beta.5 is released.
